addons/hw_l10n_eg_eta/controllers/main.py

```diff
@@ -0,0 +1,170 @@
+        return crypt_context.verify(token, stored_hash)
+
+    @http.route('/hw_l10n_eg_eta/certificate', type='http', auth='none', cors='*', csrf=False, save_session=False)
+    def eta_certificate(self, pin, secret):
+        """
+        Gets the certificate from the token and returns it to the main odoo instance so that we can prepare the
+        cades-bes object on the main odoo instance rather than this middleware
+        @param pin: pin of the token
+        @param secret: secret shared with the main odoo instance
+        @return:
+        """
+        if not self._is_secret_valid(secret):
+            return self.unauthorized_error()
+        session, error = self._get_session(pin)
+        if error:
+            return error
+        try:
+            cert = session.findObjects([(PyKCS11.CKA_CLASS, PyKCS11.CKO_CERTIFICATE)])[0]
+            payload = {
+                'certificate': base64.b64encode(session.getAttributeValue(cert, [PyKCS11.VALUE])[0]).decode('utf-8')
+            }
+            return json.dumps(payload)
+        except Exception as ex:
+            return self.get_error_template(str(ex), '')
+        finally:
+            session.logout()
+            session.closeSession()
+
+    @http.route('/hw_l10n_eg_eta/sign', type='http', auth='none', cors='*', csrf=False, save_session=False)
+    def eta_sign(self, pin, secret, invoices):
+        """
+        Gets cades-bes binaries from the odoo server for each invoice, signs each of them individually using the token
+        then returns the signatures
+        @param pin: pin of the token
+        @param secret: secret shared with the main odoo instance
+        @param invoices: dictionary of invoices. Keys are invoices ids, value are the base64 encoded binaries to sign
+        """
+        if not self._is_secret_valid(secret):
+            return self.unauthorized_error()
+        session, error = self._get_session(pin)
+        if error:
+            return error
+        try:
+            cert = session.findObjects([(PyKCS11.CKA_CLASS, PyKCS11.CKO_CERTIFICATE)])[0]
+            cert_id = session.getAttributeValue(cert, [PyKCS11.CKA_ID])[0]
+            priv_key = session.findObjects([(PyKCS11.CKA_CLASS, PyKCS11.CKO_PRIVATE_KEY), (PyKCS11.CKA_ID, cert_id)])
+
+            invoice_dict = dict()
+            invoices = json.loads(invoices)
+            for invoice, eta_inv in invoices.items():
+                to_sign = base64.b64decode(eta_inv)
+                signed_data = session.sign(priv_key, to_sign, PyKCS11.Mechanism(PyKCS11.CKM_SHA256_RSA_PKCS, None))
+                invoice_dict[invoice] = base64.b64encode(signed_data).decode('utf-8')
+
+            payload = {
+                'invoices': json.dumps(invoice_dict),
+            }
+            return json.dumps(payload)
+        except Exception as ex:
+            return self.get_error_template(str(ex), '')
+        finally:
+            session.logout()
+            session.closeSession()
```

Comment:
```suggestion
            
        except Exception as ex:
            return self.get_error_template(str(ex), '')
        finally:
            session.logout()
            session.closeSession()
        payload = {
                'invoices': json.dumps(invoice_dict),
            }
        return json.dumps(payload)
```

Not sure about this actually: the return in the try..except is actually pretty cool.  It is just that it is good to encapsulate more specific errors.